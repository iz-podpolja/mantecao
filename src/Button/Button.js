import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { darken } from 'polished'

import { colors } from '../variables'
import { media } from '../utils/style'

const getHoverColor = props => {
  return props.borderOnly ? props.color : darken(0.1, props.color)
}

const StyledButton = styled.button`
  background-color: ${props => (props.borderOnly ? colors.white : props.color)};
  border: 1px solid ${props => props.color};
  border-radius: 4px;
  color: ${props => (props.borderOnly ? props.color : colors.white)};
  cursor: pointer;
  font-size: 14px;
  padding: 11px 22px;
  position: relative;
  white-space: normal;

  &:hover {
    background-color: ${props => getHoverColor(props)};
    border-color: ${props => getHoverColor(props)};
    color: ${colors.white};
  }

  &:focus {
    background-color: ${props => getHoverColor(props)};
    border-color: ${props => getHoverColor(props)};
    color: ${colors.white};
  }

  &:disabled {
    cursor: not-allowed;
    opacity: 0.5;

    &:hover {
      background-color: ${props =>
        props.borderOnly ? colors.white : darken(0.1, props.color)};
      border-color: ${props => getHoverColor(props)};
      color: ${props => (props.borderOnly ? props.color : props.white)};
    }
  }

  ${media.medium`
    font-size: 16px;
    padding: 11px 30px;
  `}
`

const Button = props => <StyledButton {...props} />

Button.defaultProps = {
  color: colors.blueBayoux,
  borderOnly: false
}

Button.propTypes = {
  color: PropTypes.string,
  borderOnly: PropTypes.bool
}

export default Button
