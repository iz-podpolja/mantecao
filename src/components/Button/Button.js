import React, {Component, PropTypes} from 'react'

import {colors} from '../../variables'
import {colorLuminance} from '../../utils/colorUtils'

const defaultStyle = {
  padding: '0.8rem 1.5rem',
  color: colors.white,
  backgroundColor: colors.alto,
  fontSize: '1.5rem',
  outline: 'none',
  cursor: 'pointer',
  border: 'none',
  borderRadius: '4px',
  boxSizing: 'border-box',
  verticalAlign: 'middle'
}

/**
 * Normal Button - onClick and disabled props, and can have styles applied
 */
class Button extends Component {
  constructor (props) {
    super(props)
    let componentStyle = this._getComponentStyle(props.style)
    let hoverStyle = this._getHoverStyle(componentStyle)
    let disabledStyle = this._getDisabledStyle(componentStyle)
    this.state = {
      componentStyle,
      hoverStyle,
      disabledStyle,
      hover: false
    }
  }
  componentWillReceiveProps (nextProps) {
    if (nextProps.style || nextProps.disabled) {
      // recompute the styles when props dictate
      let componentStyle = this._getComponentStyle(nextProps.style)
      let hoverStyle = this._getHoverStyle(componentStyle)
      let disabledStyle = this._getDisabledStyle(componentStyle)
      this.setState({
        componentStyle,
        hoverStyle,
        disabledStyle
      })
    }
  }
  render () {
    let style = {}
    if (this.props.disabled) {
      style = this.state.disabledStyle
    } else if (this.state.hover) {
      style = this.state.hoverStyle
    } else {
      style = this.state.componentStyle
    }
    return (
      <button
        type='button'
        className={this.props.className}
        style={style}
        onMouseEnter={this._onMouseEnter}
        onMouseLeave={this._onMouseLeave}
        disabled={this.props.disabled}
        onClick={this.props.onClick}
      >
        {this.props.children}
      </button>
    )
  }
  _onMouseEnter = () => {
    this.setState({hover: true})
  }
  _onMouseLeave = () => {
    this.setState({hover: false})
  }
  _getComponentStyle = (customStyle = {}) => {
    const {textColor, bgColor, borderColor} = this.props
    return Object.assign({},
      defaultStyle,
      customStyle,
      textColor ? {color: textColor} : {},
      bgColor ? {backgroundColor: bgColor} : {},
      borderColor ? {border: '1px solid ' + borderColor} : {}
    )
  }
  _getHoverStyle = (componentStyle) => {
    const {hoverTextColor, hoverBgColor, hoverBorderColor} = this.props
    return Object.assign({},
      componentStyle,
      {backgroundColor: colorLuminance(componentStyle.backgroundColor, -0.1)},
      hoverTextColor ? {color: hoverTextColor} : {},
      hoverBgColor ? {backgroundColor: hoverBgColor} : {},
      hoverBorderColor ? {border: '1px solid ' + hoverBorderColor} : {}
    )
  }
  _getDisabledStyle = (componentStyle) => {
    return Object.assign({},
      componentStyle,
      {opacity: 0.4, cursor: 'not-allowed'}
    )
  }
}

Button.propTypes = {
  bgColor: PropTypes.string,
  borderColor: PropTypes.string,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  hoverBgColor: PropTypes.string,
  hoverBorderColor: PropTypes.string,
  hoverTextColor: PropTypes.string,
  onClick: PropTypes.func,
  style: PropTypes.object,
  textColor: PropTypes.string
}

export default Button
