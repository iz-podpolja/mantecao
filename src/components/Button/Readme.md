Example:

    <Button>a button</Button>
