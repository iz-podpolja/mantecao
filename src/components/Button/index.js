import React, {Component, PropTypes} from 'react'
import styled from 'styled-components';

import {colors} from '../../variables'
import {colorLuminance} from '../../utils/colorUtils'

const Button = styled.button`
  padding: 0.8rem 2rem;
  color: ${props => props.color || colors.white};
  background-color: ${props => props.background || colors.alto};
  font-size: 1.5rem;
  outline: none;
  cursor: pointer;
  border: ${props => props.border ? '3px solid ' + props.border : 'none'};
  border-radius: 4px;
  box-sizing: border-box;
  vertical-align: middle;

  &:hover {
    color: ${props => props.background || colors.alto};
    background-color: ${props => props.background ? colorLuminance(props.background, -0.5) : colors.white};
  }
`

Button.propTypes = {
  color: PropTypes.string,
  background: PropTypes.string,
  border: PropTypes.string
}

export default Button
