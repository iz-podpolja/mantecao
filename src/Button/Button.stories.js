import React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'
import Button from './Button'

storiesOf('Button', module)
  .add('default', () => <Button>yo</Button>)
  .add('with color', () => <Button color="red">yo color</Button>)
  .add('border only', () => <Button borderOnly>yo</Button>)
  .add('border only with color', () =>
    <Button borderOnly color="red">yo color</Button>
  )
  .add('diabled', () => <Button disabled>yo</Button>)
  .add('diabled border only', () => <Button borderOnly disabled>yo</Button>)
