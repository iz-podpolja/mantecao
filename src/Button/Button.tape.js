import React from 'react'
import test from 'tape'
import { spy } from 'sinon'
import { shallow, mount } from 'enzyme'
import { setupDOM } from '../utils/testUtils'
setupDOM()

import { hexToRgb } from '../utils/colorUtils'

import Button from './Button'

const testProps = {
  textColor: '#ABCDEF',
  hoverTextColor: '#FEDCBA',
  bgColor: '#123456',
  hoverBgColor: '#654321',
  borderColor: '#000000',
  hoverBorderColor: '#FFFFFF'
}

const testChildren = <h1 className="unique">I am a child</h1>

test('Button contains a <button>', function(t) {
  const component = shallow(<Button>{testChildren}</Button>)
  t.equal(component.find('button').length, 1)
  t.end()
})

test('Button renders children', function(t) {
  const component = shallow(<Button>{testChildren}</Button>)
  t.equal(component.contains(testChildren), true)
  t.end()
})

test('Button fires onClick events', function(t) {
  const onClick = spy()
  const component = mount(<Button onClick={onClick}>{testChildren}</Button>)
  component.find('button').simulate('click')
  t.equal(onClick.calledOnce, true)
  t.end()
})

test('Button allows props to be set', function(t) {
  const component = mount(
    <Button style={testProps.style}>
      {testChildren}
    </Button>
  )
  t.deepEqual(component.props().style, testProps.style)
  t.end()
})

test('Button changes style on hover with assigned props', function(t) {
  const onMouseEnter = spy()
  const onMouseLeave = spy()
  const component = mount(
    <Button bgColor={testProps.bgColor} hoverBgColor={testProps.hoverBgColor}>
      {testChildren}
    </Button>
  )
  const htmlStyle = 'background-color: ' + hexToRgb(testProps.bgColor) + ';'
  const htmlHoverStyle =
    'background-color: ' + hexToRgb(testProps.hoverBgColor) + ';'
  t.notEqual(component.find('button').html().indexOf(htmlStyle), -1)
  component.find('button').simulate('mouseEnter')
  t.notEqual(component.find('button').html().indexOf(htmlHoverStyle), -1)
  component.find('button').simulate('mouseLeave')
  t.notEqual(component.find('button').html().indexOf(htmlStyle), -1)

  t.end()
})

test('Button has default styles', function(t) {
  const component = mount(<Button>{testChildren}</Button>)
  const htmlStyle = 'background-color: ' + hexToRgb('#DDDDDD') + ';'
  t.notEqual(component.find('button').html().indexOf(htmlStyle), -1)
  t.end()
})

test('Button allows style changes via "style" prop', function(t) {
  const component = mount(
    <Button style={{ borderRadius: '80px' }}>
      {testChildren}
    </Button>
  )
  const htmlStyle = 'border-radius: 80px;'
  t.notEqual(component.find('button').html().indexOf(htmlStyle), -1)
  t.end()
})

test('Button allows style changes via special props', function(t) {
  const component = mount(
    <Button
      textColor={testProps.textColor}
      bgColor={testProps.bgColor}
      borderColor={testProps.borderColor}
    >
      {testChildren}
    </Button>
  )
  const textColor = 'color: ' + hexToRgb(testProps.textColor) + ';'
  t.notEqual(component.find('button').html(), -1)

  const bgColor = 'background-color: ' + hexToRgb(testProps.bgColor) + ';'
  t.notEqual(component.find('button').html(), -1)

  const borderColor =
    'border: 1px solid ' + hexToRgb(testProps.borderColor) + ';'
  t.notEqual(component.find('button').html(), -1)
  t.end()
})
