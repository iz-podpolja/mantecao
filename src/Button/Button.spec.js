import React from 'react'
import renderer from 'react-test-renderer'
import Button from './Button'

const ON_CLICK_ACTION = jest.fn()

describe('Button Component', () => {
  it('renders properly with color prop', () => {
    const tree = renderer
      .create(<Button color="lemonchiffon">whatup</Button>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders properly with hollow prop', () => {
    const tree = renderer.create(<Button hollow>whatup</Button>).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
