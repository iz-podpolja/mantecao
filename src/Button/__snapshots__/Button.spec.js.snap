// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component renders properly with color prop 1`] = `
<button
  className="sc-bdVaJa iVPjYN"
  color="lemonchiffon"
>
  whatup
</button>
`;

exports[`Button Component renders properly with hollow prop 1`] = `
<button
  className="sc-bdVaJa bbdLtm"
  color="#556285"
>
  whatup
</button>
`;
